python/docs/source/user_guide/sql/python_udtf.rst

```diff
@@ -0,0 +1,216 @@
+Python User-defined Table Functions (UDTFs)
+===========================================
+
+Spark 3.5 introduces Python user-defined table functions (UDTFs), a new type of user-defined function. 
+Unlike scalar functions that return a single result value for every input, UDTFs is invoked in the ``FROM``
+clause of a query and returns an entire relation as output.
+Each UDTF call can accept zero or more arguments.
+These arguments can either be scalar expressions or table arguments that represent entire input relations.
+
+Implementing a Python UDTF
+--------------------------
+
+.. currentmodule:: pyspark.sql.functions
+
+To implement a Python UDTF, you can define a class implementing the methods:
+
+.. code-block:: python
+
+    class PythonUDTF:
+
+        def __init__(self) -> None:
+            """
+            Initialize the user-defined table function (UDTF).
+
+            This method serves as the default constructor and is called once when the
+            UDTF is instantiated on the executor side.
+            
+            Any class fields assigned in this method will be available for subsequent
+            calls to the `eval` and `terminate` methods.
+
+            Notes
+            -----
+            - This method does not accept any extra arguments.
+            - You cannot create or reference the Spark session within the UDTF. Any
+              attempt to do so will result in a serialization error.
+            """
+            ...
+
+        def eval(self, *args: Any) -> Iterator[Any]:
+            """
+            Evaluate the function using the given input arguments.
+
+            This method is required and must be implemented.
+
+            Argument Mapping:
+            - Each provided scalar expression maps to exactly one value in the
+              `*args` list.
+            - Each provided table argument maps to a pyspark.sql.Row object containing
+              the columns in the order they appear in the provided input relation.
+
+            This method is called on every input row, and can produce zero or more
+            output rows. Each element in the output tuple corresponds to one column
+            specified in the return type of the UDTF.
+
+            Parameters
+            ----------
+            *args : Any
+                Arbitrary positional arguments representing the input to the UDTF.
+
+            Yields
+            ------
+            tuple
+                A tuple representing a single row in the UDTF result relation.
+                Yield as many times as needed to produce multiple rows.
+
+            Notes
+            -----
+            - The result of the function must be a tuple representing a single row
+              in the UDTF result relation.
+            - UDTFs currently do not accept keyword arguments during the function call.
+
+            Examples
+            --------
+            >>> def eval(self, x: int, y: int) -> Iterator[Any]:
```

Comment:
This example does not use `*args` but accepts the exact arguments as a list instead :) can we mention this is possible in the above description as well, and add an example with `*args*` and `**kwargs` as well, since all three ways are supported?

Reply:
I was thinking about this, but then I realized that using `*args` is not as readable as using variable names. It might not be something we should encourage users to use. For example:
```
def eval(self, *args):
    yield args[0], args[1]
```